db/fetcher/fetcher.go

```diff
@@ -218,91 +264,166 @@ func (df *DocumentFetcher) startNextSpan(ctx context.Context) (bool, error) {
-func (df *DocumentFetcher) nextKey(ctx context.Context) (spanDone bool, err error) {
-	// get the next kv from nextKV()
-	spanDone, df.kv, err = df.nextKV()
-	// handle any internal errors
-	if err != nil {
-		return false, err
+// It returns true if the current doc is completed.
+// The first call to nextKey CANNOT have seekNext be true (ErrFailedToSeek)
+func (df *DocumentFetcher) nextKey(ctx context.Context, seekNext bool) (spanDone bool, docDone bool, err error) {
+	// safety against seekNext on first call
+	if seekNext && df.kv == nil {
+		return false, false, ErrFailedToSeek
+	}
+
+	if seekNext {
+		fmt.Println("SEEKING...")
+		curKey := df.kv.Key
+		curKey.FieldId = "" // clear field so prefixEnd applies to dockey
+		seekKey := curKey.PrefixEnd().ToString()
+		fmt.Printf("seeking from %s to %s\n", curKey.ToString(), seekKey)
+		spanDone, df.kv, err = df.seekKV(seekKey)
+		// handle any internal errors
+		if err != nil {
+			return false, false, err
+		}
+	} else {
+		fmt.Println("NEXT...")
+		spanDone, df.kv, err = df.nextKV()
+		// handle any internal errors
+		if err != nil {
+			return false, false, err
+		}
 	}
+
 	if df.kv != nil && (df.kv.Key.InstanceType != core.ValueKey && df.kv.Key.InstanceType != core.DeletedKey) {
 		// We can only ready value values, if we escape the collection's value keys
 		// then we must be done and can stop reading
 		spanDone = true
 	}
 
+	if df.kv != nil {
+		fmt.Printf("key: %v | value: %v\n", df.kv.Key.ToString(), df.kv.Value)
+	} else {
+		fmt.Println("NO KV")
+	}
+
 	df.kvEnd = spanDone
 	if df.kvEnd {
-		_, err := df.startNextSpan(ctx)
+		moreSpans, err := df.startNextSpan(ctx)
 		if err != nil {
-			return false, err
+			return false, false, err
 		}
-		return true, nil
+		return !moreSpans, true, nil
 	}
 
 	// check if we've crossed document boundries
-	if df.doc.Key != nil && df.kv.Key.DocKey != string(df.doc.Key) {
+	if (df.doc.Key != nil && df.kv.Key.DocKey != string(df.doc.Key)) || seekNext {
 		df.isReadingDocument = false
-		return true, nil
+		return false, true, nil
 	}
-	return false, nil
+	return false, false, nil
 }
 
 // nextKV is a lower-level utility compared to nextKey. The differences are as follows:
 // - It directly interacts with the KVIterator.
 // - Returns true if the entire iterator/span is exhausted
 // - Returns a kv pair instead of internally updating
-func (df *DocumentFetcher) nextKV() (iterDone bool, kv *core.KeyValue, err error) {
+func (df *DocumentFetcher) nextKV() (iterDone bool, kv *KeyValue, err error) {
+	done, dsKey, res, err := df.nextKVRaw()
+	if done || err != nil {
+		return done, nil, err
+	}
+
+	kv = &KeyValue{
+		Key:   dsKey,
+		Value: res.Value,
+	}
+	return false, kv, nil
+}
+
+// seekKV will seek through results/iterator until it reaches
+// the target key, or if the target key doesn't exist, the
+// next smallest key that is greater than the target.
+func (df *DocumentFetcher) seekKV(key string) (bool, *KeyValue, error) {
```

Comment:
todo: It doesn't look like `key` needs to be a string, and it feels very strange that it is. Is this just because there is no easy to use std Golang []byte comparator and you want to save immediate dev time? If so I think it is worth documenting that (and very-maybe an issue).

Reply:
This was mostly just a conceptual similarity choice as atm the KV level keys are considered strings, but digging further into it theres benefits to being a string.

Firstly, there is an equivalent byte comparator as string comparator.

But, at the moment, `core.DataStoreKey.Bytes()` just calls `[]byte(datastorekey.ToString())` anyway. Which will result in an extra allocation anyway ("safe" conversion between strings and bytes allocates+copies).

At the moment, theres just a lot of conversion back and forth between strings, bytes, back to strings, and bytes, and prob a few more. So until we get organized around the proper binary serialization, it doesn't really matter which to use here. But, the one fact we know is that `datastorekey.ToString()` is cheaper than `datastorekey.Bytes()`. So its a toss.

Reply:
I think we can revisit this when we change the key encoding.

Reply:
> At the moment, theres just a lot of conversion back and forth between strings, bytes, back to strings, and bytes, and prob a few more. So until we get organized around the proper binary serialization, it doesn't really matter which to use here. But, the one fact we know is that datastorekey.ToString() is cheaper than datastorekey.Bytes(). So its a toss.

Ah fair enough, happy as-is then.

> I think we can revisit this when we change the key encoding.

I do see that as quite tangential/separate, and wouldnt recommend doing those two tasks together.